### PR TITLE
Reference support

### PR DESCRIPTION
**Pull request by mattbaker**
Hey @JakeBecker, love ElixirLS and the VSCode plugin. I wanted to take a crack at reference support (https://github.com/JakeBecker/elixir-ls/issues/19).

This seems to handle modules and functions pretty well, but there's no support for variables since I'm just leveraging `Mix.Tasks.Xref`.

Here's a little GIF:

refs

Figured I'd get it front of you to see if the direction works for you. Happy to add tests and refactor if this seems like it could actually be useful.

CC @balduncle as an interested party

### CODE REVIEW

**Comment by webdeb** [comment]
AFAIK, it should be also possible to make edits in the references panel, right?

**Comment by mattbaker** [comment]
@webdeb — hmm, are you sure about that? 

Either way, I don't _think_ it's up to the language server anyway. The protocol dictates that we just send back some file/location data for each reference and the UI is up to the editor.

**Comment by msaraiva** [comment]
@mattbaker This is probably a great feature to include in ElixirSense itself. The `references.ex` really looks like it belongs there. This way other editors could also make use of your feature. What do you think?

**Comment by mattbaker** [comment]
@msaraiva — For sure, no objection here, it's using a lot from ElixirSense already. Time is about to get tight on my end so it may be a moment before I can move it over and PR it into ElixirSense.

@JakeBecker — what do you think? Would you prefer to wait until I can get it into ElixirSense?

**Comment by JakeBecker** [comment]
Having it in Elixir Sense sounds good to me, but I'm happy to proceed for now and refactor later once it's in Elixir Sense. I'll likely review properly and try to write a test and merge tomorrow.

You might also consider trying to contribute some changes to Xref in Elixir itself so in future Elixir versions we can avoid having to capture and parse Xref's output. That could be useful in other feature providers in ElixirLS, and the Elixir guys have seemed pretty open to ideas like that so far.

**Comment by JakeBecker** [comment]
PS. @msaraiva thanks for making Elixir Sense, it's been super useful everyone in ElixirLS 👍

**Comment by mattbaker** [comment]
> Having it in Elixir Sense sounds good to me, but I'm happy to proceed for now and refactor later once it's in Elixir Sense. I'll likely review properly and try to write a test and merge tomorrow.

Cool, I'd be happy to contribute that refactor once it's into ElixirSense too. Also would love to be tagged when you write the test so I can see how you approach it. I realize this is really only half a PR without test coverage. 

> You might also consider trying to contribute some changes to Xref in Elixir itself so in future Elixir versions we can avoid having to capture and parse Xref's output.

_Definitely_, it was on my todo list. Glancing at the source, I think it'd be a small change provided they're open to adding a function to the task we can call. Fully aware of how hacky it is to capture stdout :D

> PS. @msaraiva thanks for making Elixir Sense, it's been super useful everywhere in ElixirLS 👍

+1 to that, I honestly wouldn't have been able to throw this together so quickly with ElixirSense. Consequently, I'd love to give back and contribute this in as soon as time allows @msaraiva.

**Comment by msaraiva** [comment]
> For sure, no objection here, it's using a lot from ElixirSense already. Time is about to get tight on my end so it may be a moment before I can move it over and PR it into ElixirSense.

@mattbaker If don't mind, I could try to move `references.ex` to ElixirSense and create the tests today. I'll point out in the README that it was your contribution.

> thanks for making Elixir Sense, it's been super useful everywhere in ElixirLS

@JakeBecker I'm glad it's been useful. I also need to thank you for creating ElixirLS for VS Code. It is my official editor now.

BTW. Since I'm not using Atom anymore, I'll try to find someone else to maintain the atom-elixir package. The CS/JS part was consuming a lot of my time. I'll focus now on ElxirSense. This way I can deploy more fixes and features much faster.

**Comment by mattbaker** [comment]
@msaraiva — go for it, by all means. I just pushed a commit that gets rid of the regex,  I found `Mix.Utils.parse_mfa/1` which can do the validation. The `xref` mix task itself uses `parse_mfa` so it's presumably safer than that home-rolled regex ;)

**Comment by mattbaker** [comment]
@msaraiva — oops, I meant to say "I honestly wouldn't have been able to throw this together so quickly _without_ ElixirSense", it was meant to be a compliment but I typo'd >_<

**Comment by mattbaker** [comment]
@JakeBecker — Threw out the proposal on the elixirdev list: https://groups.google.com/forum/#!topic/elixir-lang-core/DwbVKBH0goE

**Comment by mattbaker** [comment]
José pointed me in the direction of `calls/0`, combined with `Mix.Utils.parse_mfa/1` I was able to vastly simplify this PR. No more capturing of stdout, regexes, etc :)

Feel silly I passed over it to begin with, but I'm glad it exists!

**Comment by mattbaker** [comment]
Edge case not yet covered — finding references to alias modules won't work. 

Hopefully that wouldn't be too hard to add, I think I saw some references to expanding modules with aliases in ElixirSense, @msaraiva?

**Comment by msaraiva** [comment]
@mattbaker yes, you can retrieve the actual module + function from the subject using `ElixirSense.Core.Introspection.actual_mod_fun/4` like [this](https://github.com/msaraiva/elixir_sense/blob/43874b76f7ed8491d7350b4532613e7f601f7580/lib/elixir_sense/providers/definition.ex#L24).

**Comment by JakeBecker** [comment]
Good find on `Mix.Tasks.Xref.calls/0`, I hadn't known about that function. I ended up unexpectedly busy today, will review tomorrow instead. In the meantime, only thing I have to add is that the docs for `Mix.Tasks.Xref.calls/0` say that it will not work for umbrella apps, but it should be pretty straightforward to handle this case, something like the following:

```
if Mix.Project.umbrella?() do
  Enum.flat_map(Mix.Project.apps_paths(), fn {app, path} ->                                                  
    Mix.Project.in_project(app, path, fn _ -> Mix.Tasks.Xref.calls() end)
  end)
end
```

**Comment by JakeBecker** [comment]
Though, now that I think about it, there's a risk of this getting wonky if it happens concurrently with compilation. Will think on that some more.

**Comment by mattbaker** [comment]
@JakeBecker — Oh I didn't see that, good catch, I'm in umbrella projects most of the time at work so it's certainly relevant to me.

As an aside, I've updated the PR to use more of ElixirSense as I've learned more about it. I'd say it's simpler now, and it should support module aliases also thanks to ES.

@msaraiva — now it looks even more like it belongs in ElixirSense than before 😆

**Comment by mattbaker** [comment]
@JakeBecker — curious to hear more about the concurrent compilation, how does that work? Can we add a mutex that will delay the xref until compilation is done?

File: apps/language_server/lib/language_server/providers/references.ex [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  defp line_environment(text, line) do
+    Parser.parse_string(text, true, true, line + 1) |> Metadata.get_env(line + 1)
+  end
+
+  defp subject_at_cursor(text, line, character) do
+    Source.subject(text, line + 1, character + 1)
+  end
+
+  defp expand_mod_fun(nil, _environment), do: nil
```
Comment (JakeBecker) [inline apps/language_server/lib/language_server/providers/references.ex]: Dialyzer says this case isn't necessary: `The pattern <'nil', _environment@1> can never match the type <{atom(),atom()},_>`

File: apps/language_server/lib/language_server/server.ex [hunk 2]
```diff
@@ -391,6 +403,7 @@ defmodule ElixirLS.LanguageServer.Server do
       "hoverProvider" => true,
       "completionProvider" => %{"triggerCharacters" => ["."]},
       "definitionProvider" => true,
+      "referencesProvider" => true,
```
Comment (JakeBecker) [inline apps/language_server/lib/language_server/server.ex]: The `Mix.Tasks.Xref.calls/0` function was added in 1.6, so we should add `References.supported?/0` function we call here like in the `Formatting` provider.

**Comment by JakeBecker** [comment]
@mattbaker The language server runs builds in a separate process, so it can happen concurrently with the request providers. Using `Mix.Project.in_project` changes both the project stack and the cwd, both of which will almost certainly screw with any compilation happening at the same time. I don't at the moment know how we should handle that, but it's fine if it initially only works in non-umbrella projects.

**Comment by mattbaker** [comment]
@JakeBecker — I made a couple updates upon your suggestion. The precise technique you used to check for the existence of the exported function for the formatter didn't work for me, it would only work after (I think) I caused the module to be loaded. I think what I added should get the job done though.

Re compilation: I agree umbrella support could be a separate PR. Is `ElixirLS.LanguageServer.Build` where the compilation is kicking off? I wonder if we could just have the build [acquire a lock](http://erlang.org/doc/man/global.html) that would prevent our `xref` code from running during compilation. The lock/transaction stuff in the erlang global module makes it incredibly simple, I used it recently to solve some minor concurrency issues that didn't justify a larger redesign.

Maybe we could even have the reference request display something helpful to the user if they attempt to Find References during compilation, it'll error out if it can't acquire the lock after a few retries so we'd be able to detect the issue.

Anyway, just a thought I had on the way home.

**Comment by JakeBecker** [comment]
Merged!

Using locks struck me initially as a not-very-Elixir way of handling it, but looking at it more, it does seem like global locks are a clean way to keep the global state from causing problems, so I'm not opposed. The build stuff is mostly in `ElixirLS.LanguageServer.Build`, and it's called from `ElixirLS.LanguageServer.Server.trigger_build/1`

**Comment by mattbaker** [comment]
Agreed, locks are not usually the first thing I reach for when dealing with concurrency issues, but it seems like it might be appropriate in this context. I think your point is a good one, which is that the status of that build is something of a global state.

I'll own a new issue for this.

**Pull request merged**